[refactor] #308 - 홈 조회에서의 schedule 데이터의 반복 쿼리 개선 완료

move-hoon opened the pull request:
## Related issue 🛠

<!-- 관련 이슈 번호를 적어주세요 -->
- closes #308 

## Work Description ✏️

<!-- 작업 내용을 간단히 소개주세요 -->

- 홈 조회에서 minDueDate 조회를 위해 발생했던 N번의 반복쿼리를 1번의 쿼리로 계산하도록 변경했습니다.
- DB단에서 dueDate를 산정할 PerformanceDate를 결정하는 쿼리가 복잡하다 보니, QueryDSL을 사용했습니다.
- 현재 로컬에서 성능 테스트를 해본 결과 응답속도 부분에서는 개선이 없어서 추후 인덱스를 적용하거나 dueDate 집계컬럼을 따로 생성해서 배치 작업을 할 예정입니다.


## Trouble Shooting ⚽️

<!-- 어떤 위험이나 장애를 발견했는지 적어주세요 -->

## Related ScreenShot 📷

<!-- 관련 스크린샷을 첨부해주세요 -->

### AS-IS
```sql
Hibernate: 
    select
        s1_0.id,
        s1_0.is_booking,
        s1_0.performance_id,
        s1_0.performance_date,
        s1_0.schedule_number,
        s1_0.sold_ticket_count,
        s1_0.total_ticket_count 
    from
        schedule s1_0 
    where
        s1_0.performance_id=?
```

- 하나의 performance에 연관된 여러 schedule 객체를 조회해야 했기에 위의 쿼리가 N번 나가는 상황이 발생했었습니다.

### TO-BE
```sql
Hibernate: 
    select
        s1_0.performance_id,
        case 
            when (min(case 
                when (s1_0.performance_date>=?) 
                    then s1_0.performance_date 
                else null 
        end) is not null) 
            then min(case 
                when (s1_0.performance_date>=?) 
                    then s1_0.performance_date 
                else null 
        end) 
    else min(case 
        when (s1_0.performance_date<?) 
            then s1_0.performance_date 
        else null 
end) 
end 
from
    schedule s1_0 
where
    s1_0.performance_id in (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?) 
group by
    s1_0.performance_id
```
- 1번의 쿼리로 minDueDate 까지 판단해서 가져올 수 있도록 개선하였습니다.

## Uncompleted Tasks 😅

<!-- 끝내지 못한 작업을 적어주세요 -->

## To Reviewers 📢

<!-- 리뷰어들에게 물어볼 점, 할 말 등을 적어주세요 -->
- 이번에 QueryDSL를 사용하면서 dao단에서 조회할 때  DTO로 반환하는 로직을 사용했습니다.
- 해당 방법에는 4가지 방법이 있는데(bean, field, constructor, @queryfrojection) @queryfrojection을 선택하게 되었습니다.
- 그 이유는 아래 링크를 보고 공감이 되어서 선택을 하게 되었으니 한번 읽어보시면 도움이 될 것 같습니다.

[QueryDSL 참고링크](https://developer-nyong.tistory.com/65#article-3-2--projections-fields
)

move-hoon committed:
[#308] chore(HomeController): 변수 네이밍 변경

move-hoon committed:
[#308] fix: findAllByPerformanceId를 사용하도록 통일

move-hoon committed:
[#308] refactor(ServiceLoggingAspect): 서비스 반환 로그 debug 레벨로 변경

- 데이터가 늘어날 수록 로그가 방대하게 늘어나는 문제로 인해 debug 레벨로 변경

move-hoon committed:
[#308] refactor: 패키지 이동

move-hoon committed:
[#308] chore: 패키지 이동으로 인한 import문 변동

move-hoon committed:
[#308] feat: 1번의 쿼리로 여러 개의 performance_id의 최소 performance_date를 가져오도록 로직 구현

move-hoon committed:
[#308] feat(MinPerformanceDateDto) 특정 공연의 최소 공연일 조회를 위한 MinPerformanceDateDto 추가

- QueryDSL의 `@QueryProjection`을 활용하여 생성자를 통해 필드를 초기화할 수 있도록 설계

move-hoon committed:
[#308] feat(MinPerformanceDateResponse): 공연 ID와 최소 공연 날짜를 매핑한 `performanceDateMap`을 담는 DTO 생성

move-hoon committed:
[#308] feat(ScheduleService): 주어진 공연 ID 목록에 대한 최소 공연 날짜를 조회하는 메서드 구현

move-hoon committed:
[#308] refactor(HomeService): 홈 조회 메서드 리팩토링

move-hoon committed:
[#308] fix(ScheduleRepositoryCustomImpl): 변수 네이밍 수정

hyerinhwang-sailin reviewed (commented):
저도 글을 읽어보니 repository에서 dto로 반환하는 방법 중 @queryfrojection이 제일 좋은 것 같습니다!
다만 생성된 쿼리를 봤을 때 CASE WHEN이 중첩되기도 하고 CASE WHEN 안에서 MIN()을 사용하고 있어서 추후 인덱스 적용 시 쿼리 최적화가 어려울 수 있을 같다는 생각이 들었습니다.
그래서 ORDER BY + LIMIT 방식을 사용하고, 미래 스케줄과 과거 스케줄을 분리해서 조회하면 더 좋을 것 같은데 어떻게 생각하시나요?
```
(
    -- 가장 가까운 미래 dueDate 조회
    SELECT performance_id, dueDate
    FROM schedule
    WHERE dueDate >= CURRENT_TIMESTAMP
    ORDER BY dueDate ASC
    LIMIT 1
)
UNION ALL
(
    -- 미래 dueDate가 없을 경우, 가장 가까운 과거 dueDate 조회
    SELECT performance_id, dueDate
    FROM schedule
    WHERE dueDate < CURRENT_TIMESTAMP
    ORDER BY dueDate DESC
    LIMIT 1
)
LIMIT 1;
```
이렇게 SQL 쿼리 나가면 좋을 것 같아서요!

move-hoon committed:
[#308] refactor(HomeService): 가독성 관련 리팩토링

move-hoon committed:
[#308] refactor(ScheduleRepositoryCustomImpl): 미래 일정이 존재하면 즉시 반환하도록 개선

move-hoon reviewed (commented):
```sql
SELECT performance_id, performance_date
FROM schedule
WHERE performance_id IN (...)
ORDER BY performance_date ASC
LIMIT 1;
```

이 쿼리는 모든 performanceId에 대해 단 하나의 공연 일정만 가져오는 것이기 때문에,
다음과 같은 문제가 발생하더라고요!

- LIMIT 1이 적용된 상태에서는 전체에서 가장 빠른 공연 하나만 반환
- 즉 performanceId별로 가장 가까운 일정을 가져와야 하지만, 이 방식은 하나의 공연 ID만 선택

이 밖에도 여러가지 쿼리를 도입해 보았는데, 마지막 커밋의 쿼리가 성능 및 가독성 측면에서 가장 적합하다고 생각이 들었습니다.

의견주시면 감사하겠습니다!

move-hoon committed:
[#308] chore(ScheduleRepositoryCustomImpl): 안쓰는 import문 정리

move-hoon committed:
[#308] fix(ScheduleRepositoryCustomImpl): 조기 리턴 과정 롤백

- 모든 performanceIds에 대한 minDate를 판별해야 하는데 조기 리턴을 했었고 이로 인해 계산되지 않은 performanceIds의 minDate가 존재하게 되는 문제 수정

hyerinhwang-sailin reviewed (approved):
쿼리문이 훨씬 깔끔해져서 좋네요👍
다만 단일 쿼리로 여러 공연의 duedate를 한번에 계산하는 방식이 쿼리 개수는 줄일 수 있지만 공연 수가 많아질수록 성능이 떨어질 수도 있을 것 같다는 우려도 듭니다..!! 우선 dev에서 테스트해보고 성능 모니터링을 통해 피드백 진행하면 어떨까요? 고생하셨습니다! 👏👏
